src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -524,13 +560,144 @@ class TypeComparer(initctx: Context) extends DotClass with ConstraintHandling {
+            false
+        }
+      case _ =>
+        false
+    }
+
+    /** `param2` can be instantiated to the type constructor of the LHS
+     *  or to the type constructor of one of the LHS base class instances
+     *  and the resulting type application is a supertype of `tp1`,
+     *  or fallback to fourthTry.
+     */
+    def canInstantiate(param2: PolyParam): Boolean = {
+
+      /** `param2` can be instantiated to `tycon1a`.
+       *  and the resulting type application is a supertype of `tp1`.
+       */
+      def tyconOK(tycon1a: Type) =
+        variancesConform(tycon1a.typeParams, tparams) && {
+          (ctx.mode.is(Mode.TypevarsMissContext) ||
+           tryInstantiate(param2, tycon1a.ensureHK)) &&
+          isSubType(tp1, tycon1a.appliedTo(args2))
+        }
+
+      tp1.widen match {
+        case tp1w @ HKApply(tycon1, _) =>
+          tyconOK(tycon1)
+        case tp1w =>
+          tp1w.typeSymbol.isClass && {
+            val classBounds = tycon2.classSymbols
+            def liftToBase(bcs: List[ClassSymbol]): Boolean = bcs match {
+              case bc :: bcs1 =>
+                classBounds.exists(bc.derivesFrom) && tyconOK(tp1w.baseTypeRef(bc)) ||
+                liftToBase(bcs1)
+              case _ =>
+                false
+            }
+            liftToBase(tp1w.baseClasses)
+          } ||
+          fourthTry(tp1, tp2)
+      }
+    }
+
+    /** Let `tycon2bounds` be the bounds of the RHS type constructor `tycon2`.
+     *  Let `app2 = tp2` where the type constructor of `tp2` is replaced by
+     *  `tycon2bounds.lo`.
+     *  If both bounds are the same, continue with `tp1 <:< app2`.
+     *  otherwise continue with either
+     *
+     *    tp1 <:< tp2    using fourthTry (this might instantiate params in tp1)
+     *    tp1 <:< app2   using isSubType (this might instantiate params in tp2)
+     */
+    def compareLower(tycon2bounds: TypeBounds): Boolean = {
+      val app2 = tycon2bounds.lo.applyIfParameterized(args2)
+      if (tycon2bounds.lo eq tycon2bounds.hi) isSubType(tp1, app2)
+      else either(fourthTry(tp1, tp2), isSubType(tp1, app2))
+    }
+
+    tycon2 match {
+      case param2: PolyParam =>
+        isMatchingApply(tp1) || {
+          if (canConstrain(param2)) canInstantiate(param2)
+          else compareLower(bounds(param2))
+        }
+      case tycon2: TypeRef =>
+        isMatchingApply(tp1) ||
+        compareLower(tycon2.info.bounds)
+      case tycon2: TypeVar =>
+        isSubType(tp1, tycon2.underlying.appliedTo(args2))
+      case tycon2: AnnotatedType =>
+        compareHkApply2(tp1, tp2, tycon2.underlying, args2)
+      case _ =>
+        false
+    }
+  }
+
+  /** Subtype test for the hk application `tp1 = tycon1[args1]`.
+   */
+  def compareHkApply1(tp1: Type, tycon1: Type, args1: List[Type], tp2: Type): Boolean =
+    tycon1 match {
+      case param1: PolyParam =>
+        def canInstantiate = tp2 match {
+          case AppliedType(tycon2, args2) =>
```

Comment:
This could be replaced by `case HKApply(tycon2, args2)`, no point in destructuring a simply-kinded type application